### PR TITLE
fix: make it compatible with Symfony 5 and 6

### PR DESCRIPTION
**Pull request by drupol**
This PR

- [x] Depends on #51 
- [x] Remove support of Symfony 4 and make it compatible with Symfony 5 and 6.

### CODE REVIEW

**Comment by ademarco** [comment]
Hello @drupol , unfortunately we cannot drop Symfony 4, as Drupal core still needs it.

File: src/Notification/NotificationHandler.php [hunk 1]
```diff
@@ -147,7 +147,7 @@ public function receiveNotification(ReceiveNotification $wrapper): ReceiveNotifi
                 break;
         }
 
-        $this->eventDispatcher->dispatch($event::NAME, $event);
```
Comment (ademarco) [inline src/Notification/NotificationHandler.php:150]: Good catch, I'll include this in a PR we are currently working on.

**Comment by drupol** [comment]
Ciao Marco, 

Yes, that's what I though, how about maintaining a specific branch ?

**Comment by ademarco** [comment]
What about having 4 || 5 || 6 and add a matrix in the CI to check if test pass on each of them? Is that even possible?

**Comment by drupol** [comment]
Yes, I tried that, but since the `DateTimeNormalizer` signature has been fully typed, it's not compatible between `^4` and `^5 || ^6`.

How about maintaining 2 branches of this projects?

- `1.x`  branch would be compatible with Symfony ^4
- `main` branch would be compatible with Symfony ^5 and upper

Something like that:

```mermaid
gitGraph
    commit
    commit
    branch 1.x
    checkout 1.x
    commit
    commit
    checkout main
    commit
    commit
```

**Comment by ademarco** [comment]
We could alias a class dynamically, so depending from `\Symfony\Component\HttpKernel\Kernel::VERSION` (or similar) we can alias one or the other. 

Drupal does it in tests, to have different method signatures, see below:

```php
<?php

namespace Drupal\Tests;

use Drupal\TestTools\PhpUnitCompatibility\RunnerVersion;

// In order to manage different method signatures between PHPUnit versions, we
// dynamically load a compatibility trait dependent on the PHPUnit runner
// version.
class_alias("Drupal\TestTools\PhpUnitCompatibility\PhpUnit" . RunnerVersion::getMajor() . "\TestCompatibilityTrait", 'Drupal\Tests\PhpUnitCompatibilityTrait');
```

**Comment by drupol** [comment]
True, sadly that solution exists. I don't like it much since it would require to create classes based on a condition... and that's why I was suggesting using branches. Anyway, I will try locally first and see how it goes.

In any case, I guess at some point Drupal will hopefully use Symfony 5 components.

Let me get back to you with this.

**Comment by ademarco** [comment]
Yeah, Drupal 10 will move to Symfony 6, so as soon as we migrate to that we won't need this anymore, that's why I'd prefer not having to deal with a separate branch.

**Comment by drupol** [comment]
I just pushed a complete different approach based on your feedback. I didn't think using `class_alias` was necessary here.

Let me know if it's better now.

**Comment by drupol** [comment]
Going to rebase this one now.

File: src/Serializer/Normalizer/DateTimeNormalizer4.php [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+namespace OpenEuropa\EPoetry\Serializer\Normalizer;
+
+use DateTimeInterface;
+use Symfony\Component\Serializer\Normalizer\DateTimeNormalizer as SymfonyDateTimeNormalizer;
+
+/**
+ * Extension of Symfony's DateTimeNormalize class, only for Symfony 4.
+ *
+ * @deprecated When support for Symfony 4 will be dropped, this class needs to
```
Comment (ademarco) [inline src/Serializer/Normalizer/DateTimeNormalizer4.php:13]: I'd not mark this a deprecated as this class is not a deprecated class as such, but we should keep a comment that states that this is to be used only with Symfony 4.

File: src/Serializer/Serializer.php [hunk 3]
```diff
@@ -131,7 +133,11 @@ protected function getSerializer(array $encoders = []): SymfonySerializer
             AbstractObjectNormalizer::SKIP_NULL_VALUES => true, // Null values won't be generated.
         ];
         return new SymfonySerializer([
-            new DateTimeNormalizer(),
+            // When support for Symfony 4 will be dropped, two things needs to
```
Comment (ademarco) [inline src/Serializer/Serializer.php:136]: Maybe add a `@todo` here?

**Review by ademarco — CHANGES_REQUESTED** [review]
New approach works for me, are you going to amend the test matrix too? so we can test it with Symfony 4 and 5+ at the same time?

File: src/Serializer/Serializer.php (hunk 3 above)
Comment (drupol) [reply src/Serializer/Serializer.php:136]: Oh food idea !

**Comment by drupol** [comment]
Hey,

I'm struggling a little bit with this, what would be the approach that you'd prefer to tests multiple versions of Symfony?

Thx

**Comment by ademarco** [comment]
Instead of --prefer-lowest, we could require a specific version of `symfony/http-kernel`, before running tests in the CI, which would force the other dependencies to follow up, like:
```
composer require symfony/http-kernel:~4
```
```
composer require symfony/http-kernel:~5
```
```
composer require symfony/http-kernel:~6
```

**Comment by drupol** [comment]
Ok I give up with this for today. I'm now spending more time fixing things that shouldn't happen if we would use two separate branches as suggested.

I tried to require `symfony/http-kernel:^4` locally and tests are failing:

image

I therefore removed `composer.lock` and `vendor` and require `symfony/serializer:^4` and it's also failing at the same place.

At this point, I don't know what else to do. I strongly believe all of these stuff could be avoided if we could have 2 different branches where development could happen without being forced to mix code that is not backward compatible.

**Comment by ademarco** [comment]
I've had a few tries at this and, indeed, changes are too many to keep everything on one branch. I'll create two branches, from current master, 1.x (Symfony 4) and 2.x (Symfony 6). Once 2.x is there you can try again on this starting from that branch.

**Comment by ademarco** [comment]
I have a working version for Symfony 6 on #53.

**Comment by drupol** [comment]
Hey!


This is a good news. I definitely believe that it is the way to go.

However, shouldn't it be better to branch off to `1.x` and use the `main` branch for the most recent development ?

Now going to check your new PR and give you feedback in it.


Have a good day

**Comment by drupol** [comment]
Closing in favor of #53

**Pull request closed without merging**